packages/hash/shared/src/types.ts

```diff
@@ -1,8 +1,46 @@
-// The types exported here are expected to be branded.
+import {
+  EntityId,
+  extractEntityUuidFromEntityId as extractEntityUuidFromEntityIdSubGraph,
+  extractOwnedByIdFromEntityId as extractOwnedByIdFromEntityIdSubGraph,
+  splitEntityId as splitEntityIdSubGraph,
+} from "@hashintel/hash-subgraph";
 
-export type Uuid = string;
-export type AccountId = Uuid;
-export type OwnedById = Uuid;
-export type UpdatedById = Uuid;
+type Branded<Base, Kind extends string> = {
+  /** The type of the value space that is branded */
+  readonly _base: Base;
+  /** The unique name for the branded type. */
+  readonly _kind: Kind;
+};
 
-export type EntityUuid = Uuid;
+/**
+ * The type-branding type to support nominal (name based) types.
+ */
+export type Brand<Base, Kind extends string> = Base & Branded<Base, Kind>;
+
+/** An account ID of an actor that is the owner of something */
+export type OwnedById = Brand<string, "OwnedById">;
+/** An account ID of an actor that has updated something */
+export type UpdatedById = Brand<string, "UpdatedById">;
+/** An ID to uniquely identify an account (e.g. a User or an Org) */
+export type AccountId = Brand<string, "AccountId"> | OwnedById | UpdatedById;
```

Comment:
Are you not able to do `export type OwnedById = Brand<AccountId, "OwnedById">`?

Is there a way we can rework the type to allow that possibly?, I'm guessing the issue is it would override the readonly annotations rather than wrapping them 🙃

Reply:
I spent a fair bit of time trying to allow this, and unfortunately the type always becomes `never` because of the way we compose the types right now. I can give it another shot as it would make the hierarchy a lot simpler to define

Reply:
(We collaborated offline and came up with [this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAQgTgQwHYBMIoDwwQZ2hAD2AlRygFckBrJAewHckoAfWRVdDSmhpAGig5gcAJZIA5gD4BAaTEpJUALxQA3gCgoUAPQAqXVAAqAC2ihIUWgDMowU1ABuCADbloOMAgDGZ4wmBQImQARuxoKFC62ppQcBAIKLRIziBQAPrBuBAAXLBZANwxegYm0JQiAI5uUEgIALbQVrRwtvahyOG24BAAdJHRWnEJSSnpVPK5AIJwiCBYWVCExKRsHZzcdIwCQqISigD8UADaPafYeEcARGnjqJcAurLy97lyqJKF6gC+n8WaJfZzBAALTtVBicRdCzAWiCchgMDNAJ0OpiFxQAAUtQaUEyeBQAEooRAcD1-tFCIi4AEgatUPM8E9UIsiCQUGQdhDFCpztAAGR08IMiBMhSfdTabTEqCTLxeWiUYAASQiPLCGE5EgEl1l8sVKsuknUEqltIAYrRYSpNZCBRpJVp0nicoJhBC+CbHTcJsdLhbaJcBKcejajvd7p6vsbzZblIKNW6tVA-ZbDb9Td08i1rYnbWpPVoMllcjaPQ7C7cULkrv7A8nsHA68HQ+HI+paQ24-AOhh-dqG2njZ6OwgAF5xm1QO0Fp3F127cRlqUVn011P9hCNjejpunFsRh1RkfjtU9hvbwfqeVIIRQJq0XL+uOXS5QXBQf2Fa+3zJwXKdlQXzfMgGy-JIfzHf8x2fV932wUdPm-GkECoCAnxUDECEfS0iSURRVB+K9wOQ1CAMxLCs1w-DCKQ2wUIgeC40wqDRyotRCPbej-Qxe8CUKYAuMtDFfz4zjUO4zJWM+ATSM3HicP4+iG2EzdRJkhi5Mk0SxI00d5NoNSlLHFS4EM2S9K0wogA))

Reply:
Revamped [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAQgTgQwHYBMCiAPYEkGcCWA9kgDwwK4QA0UA0vqlBFjirlAN4C+AfFALycAUFCgB6AFQSoAFQAW0UJCiEAZlGAKoANwQAbAK7RcYBAGNFchMCj52AI0SoIKKBLEiocCAhTE9IFAA+vYUEABcsGEA3J6S0vLQBkj4AI5GUEgIALbQqoRwGlqOyCguGuAQAHRuHqLevv6BQQDWDCiR9KixQlw98SIJWkoQALQlqAwA5hXKwIRQuAZgYAU2SITZDPpQABRZuVChlCgAlLMQuFWDHsyrcDYjsE5lKGRhNF2uzNio7LjAODTPiCciUJgsP7PUqYX4EYgkBiqCCFAByl2wKDB1FsSGRaIxLi+fE8AH4ouCAGTQ1CwnDw0jogEubE0JmYr5QakcADaQVxdHaAF1IskWhsAO5IXhQTyRbFcmnoSEM96UGi8-kMQWoEVQMWS6U8fpSQZQACCSBUYGARCyeigqmSZltxA0Cwmrk0ikquOZvhU6juenwZnwwACUDMFFtSBm4ZuQjuayjxABRxeAigJE5P1Y-0B0xorKVMl9eah8FKLjVOOJPF2GHlHygbVQpM67XO-D4HE83mABjgVowUAo+qQ4sIUrH7CrqDLkF6QiETwAYoQFqCXjWOFAytlCAB5ewAKwgLsiAKBcagXBoACIN4QH8aV09yIVt9W3s-H5-Xx6D8EAALyzedXneOB-1AwCVzMNMbHyQhImfLM9wPY8zwvYBIgfXBNggAA1fQjAfO9ZygZ9YjEMRUzwJDN1Qzcs09XYMIgQ8T3PS8oAfCVfAlciuFOHoEIYo4EDgZsvwzUpdmQ85x0-Gi6PE9NQmkqJZLYzTRPgxDJJA5swMEXSpKUudQNU+iNNAkzWJeXZQhA-ShHUx4EBaCA0MERtmMIbtezvWIPI0LyIE-LN-O0oLOBC9zDOACLyFMvYmyiVyBGCvp3wi58FM3fTku8gq9NiEqfM3ZzQLc2ioAAAWAXBRjuHDWrgOAClXFKpMKwKKt6uAargYqhpq1z+jopqWral0Oq6uAeu81L+v0+qZtajBIHmlFFuWyLQJGsaVqOlzRKAA)

Reply:
Added in e8af67a4df3371b721fd7a7c9ed00e40f2545246, adjusted the inner brand property names to not appear in intellisense :+1: